analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs

```diff
@@ -32,6 +34,41 @@ internal static class Binary
                 ? context.SetOperationConstraint(newConstraint)
                 : context.State;
 
+        public static ProgramState LearnBranchingConstraint(ProgramState state, IBinaryOperationWrapper binary, bool useOpposite) =>
+            binary.OperatorKind.IsAnyEquality()
+                ? LearnBranchingConstraint<ObjectConstraint>(state, binary, useOpposite) ?? LearnBranchingConstraint<BoolConstraint>(state, binary, useOpposite) ?? state
+                : state;
+
+        private static ProgramState LearnBranchingConstraint<T>(ProgramState state, IBinaryOperationWrapper binary, bool useOpposite)
+            where T : SymbolicConstraint
+        {
+            // We can fall through ?? because "constraint" and "testedSymbol" are exclusive. Symbols with the constraint will be recognized as "constraint" side.
+            if ((OperandConstraint(binary.LeftOperand) ?? OperandConstraint(binary.RightOperand)) is { } constraint
+                && (OperandSymbolWithoutConstraint(binary.LeftOperand) ?? OperandSymbolWithoutConstraint(binary.RightOperand)) is { } testedSymbol)
```

Comment:
Looks good to me as is. You may consider implementing it more like what you really want to do:
```suggestion
SymbolicConstraint constraint = null;
ISymbol testedSymbol = null;
if (OperandConstraint(binary.LeftOperand) is { } leftConstraint && OperandSymbolWithoutConstraint(binary.RightOperand) is { } rightSymbol)
{
  constraint = leftConstraint;
  testedSymbol = rightSymbol;
} else if (OperandConstraint(binary.RightOperand) is { } rightConstraint && OperandSymbolWithoutConstraint(binary.LeftOperand) is { } leftSymbol))
{
  constraint = rightConstraint;
  testedSymbol= leftSymbol;
}
if (constraint is not null && testedSymbol is not null)
```
It is much more verbose, but it is easier to read.

Reply:
I'll keep the short one. The verbose one defines that one of the `x == null` vs `null == x` situations has priority over the other.